### PR TITLE
Added tests for BinaryUtils and HttpUtils

### PR DESCRIPTION
**Pull request by freddytuxworth**
*Description of changes:*
Added unit tests for BinaryUtils and HttpUtils.

I analysed this codebase and found that these classes were not tested. I have written some unit tests for them using [Diffblue](https://www.diffblue.com/) [Cover](https://www.diffblue.com/products). Hopefully these tests will allow you to detect any regressions caused by future code changes.

Please let me know if you would like any more tests written for any other classes or repositories.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class BinaryUtilsTest {
+
+    @Rule public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void testFromHex() {
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: Would be interested to see a test for a null value passed into fromHex()

File: src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Test
+    public void testToBase64() throws Exception {
+        Assert.assertEquals("dGVzdCBzdHJpbmc=", BinaryUtils.toBase64("test string".getBytes()));
+        Assert.assertEquals("", BinaryUtils.toBase64(new byte[] {}));
+        Assert.assertEquals("AQIDBA==", BinaryUtils.toBase64(new byte[] {0x01, 0x02, 0x03, 0x04}));
+    }
+
+    @Test
+    public void testFromBase64() throws Exception {
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: Is there a way we can reach the unsupportedEncodingException?

Comment (sudo-jaa) [reply src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: Presumably pass in a base64 string that is encoded psychopathically and that'll do the trick.

File: src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        Assert.assertArrayEquals(new byte[] {}, BinaryUtils.fromBase64(""));
+        Assert.assertArrayEquals(new byte[] {0x01, 0x02, 0x03, 0x04}, BinaryUtils.fromBase64("AQIDBA=="));
+    }
+
+    @Test
+    public void testToHex() {
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: What happens if we pass a byte array of both length 8, and length > 8?

Comment (freddytuxworth) [reply src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: Tried this test case. I first got `uncaught FundamentalAxiomViolationError` but then started noticing physical objects going missing in the world around me. I'm now trapped in an alternate dimension. While it's a valid test, it seems unlikely the developers would want the same experience.

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath/", "test"));
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath/", "/test"));
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath", "/test"));
+        Assert.assertEquals("basepath/test//doubleslash", HttpUtils.appendUri("basepath", "/test//doubleslash"));
+        Assert.assertEquals("basepath/test/%2Fdoubleslash", HttpUtils.appendUri("basepath", "/test//doubleslash", true));
+    }
+
+    @Test
+    public void testIsUsingNonDefaultPort(){
+        Assert.assertFalse(HttpUtils.isUsingNonDefaultPort(URI.create("http://test:80")));
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: Add an assertFalse() for when the port is 0, and for when the port is potentially < 0.

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        Assert.assertFalse(HttpUtils.isUsingNonDefaultPort(URI.create("https://test:443")));
+        Assert.assertFalse(HttpUtils.isUsingNonDefaultPort(URI.create("scheme://other")));
+
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("http://test:9999")));
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: add an assertTrue() for when the port scheme is `http` and the port is `443`

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("https://test:9999")));
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: add an assertTrue() for when the port scheme is `https` and the port is `80`

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java (hunk 4 above)
Comment (freddytuxworth) [reply src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: https://stackoverflow.com/questions/11225373/negative-port-number-in-http-request

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        Assert.assertFalse(HttpUtils.isUsingNonDefaultPort(URI.create("scheme://other:-10")));
+
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("http://test:9999")));
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("http://test:443")));
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("https://test:9999")));
+        Assert.assertTrue(HttpUtils.isUsingNonDefaultPort(URI.create("https://test:80")));
+    }
+
+    @Test
+    public void urlEncodeInputNullTrueOutputNotNull() {
```
Comment (sudo-jaa) [inline src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: Can we run a test for some of these url replacements that contain two subsequent replaceable characters? For example: `"url+++++encode"`

File: src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java (hunk 3 above)
Comment (sudo-jaa) [reply src/test/java/com/amazon/dtasdk/utils/BinaryUtilsTest.java]: I was wondering why you weren't at your desk.

File: src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    public void testAppendUri() {
+        Assert.assertEquals("/", HttpUtils.appendUri("", ""));
+        Assert.assertEquals("basepath/two%2Bwords", HttpUtils.appendUri("basepath", "two+words"));
+        Assert.assertEquals("basepath/", HttpUtils.appendUri("basepath", null));
+        Assert.assertEquals("basepath/", HttpUtils.appendUri("basepath/", null));
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath/", "test"));
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath/", "/test"));
+        Assert.assertEquals("basepath/test", HttpUtils.appendUri("basepath", "/test"));
+        Assert.assertEquals("basepath/test//doubleslash", HttpUtils.appendUri("basepath", "/test//doubleslash"));
+        Assert.assertEquals("basepath/test/%2Fdoubleslash", HttpUtils.appendUri("basepath", "/test//doubleslash", true));
```
Comment (peterschrammel) [inline src/test/java/com/amazon/dtasdk/utils/HttpUtilsTest.java]: break lines if longer than 80 chars

**Review by peterschrammel — APPROVED** [review]

**Review by petersear — APPROVED** [review]

**Pull request closed without merging**